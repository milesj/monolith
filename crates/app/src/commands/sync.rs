use super::syncs::codeowners::SyncCodeownersArgs;
use super::syncs::hooks::SyncHooksArgs;
use crate::session::CliSession;
use clap::Subcommand;
use starbase::AppResult;
use starbase_styles::color;
use tracing::warn;

#[derive(Clone, Debug, Subcommand)]
pub enum SyncCommands {
    #[command(
        name = "codeowners",
        about = "Aggregate and sync code owners to a `CODEOWNERS` file."
    )]
    Codeowners(SyncCodeownersArgs),

    #[command(
        name = "hooks",
        about = "Generate and sync hook scripts for the workspace configured VCS."
    )]
    Hooks(SyncHooksArgs),

    #[command(
        name = "projects",
        about = "Sync all projects and configs in the workspace."
    )]
    Projects,
}

pub async fn sync(session: CliSession) -> AppResult {
    warn!(
        "This command is deprecated. Use {} instead.",
        color::shell("moon sync projects")
    );

    crate::commands::syncs::projects::sync(session).await
}
