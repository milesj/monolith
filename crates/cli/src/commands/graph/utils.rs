use super::dto::{GraphEdgeDto, GraphInfoDto, GraphNodeDto};
use crate::helpers::AnyError;
use moon_logger::info;
use moon_runner::DepGraph;
use moon_workspace::Workspace;
use petgraph::Graph;
use serde::Serialize;
use std::{collections::HashSet, env};
use tera::{Context, Tera};
use tiny_http::{Header, Response, Server};

const INDEX_HTML: &str = include_str!("graph.html.tera");
const LOG_TARGET: &str = "moon:graph::utils";

#[derive(Debug, Serialize)]
pub struct RenderContext {
    pub graph_data: String,

    pub js_url: String,
}

pub async fn setup_server() -> Result<(Server, Tera), AnyError> {
    let port = match env::var("PORT") {
        Ok(p) => p.parse::<u16>().unwrap(),
        Err(..) => 8000,
    };
    let host = match env::var("HOST") {
        Ok(h) => h,
        Err(..) => "127.0.0.1".to_string(),
    };
    let address = format!("{}:{}", host, port);
    let server = Server::http(address).unwrap();

    let tera = Tera::default();

    Ok((server, tera))
}

pub fn extract_nodes_and_edges_from_graph(graph: &Graph<String, ()>) -> GraphInfoDto {
    let edges = graph
        .raw_edges()
        .iter()
        .map(|e| GraphEdgeDto {
            source: e.source().index(),
            target: e.target().index(),
            id: format!("{}->{}", e.source().index(), e.target().index()),
        })
        .collect::<Vec<_>>();
    let mut nodes = HashSet::new();
    for edge in graph.raw_edges().iter() {
        let source = graph
            .node_weight(edge.source())
            .expect("Unable to get node")
            .clone();
        let target = graph
            .node_weight(edge.target())
            .expect("Unable to get node")
            .clone();
        nodes.insert(GraphNodeDto {
            id: edge.source().index(),
            label: source,
        });
        nodes.insert(GraphNodeDto {
            id: edge.target().index(),
            label: target,
        });
    }
    let nodes = nodes.into_iter().collect();
    GraphInfoDto { edges, nodes }
}

/// Get a serialized representation of the workspace graph.
pub async fn workspace_graph_repr(workspace: &Workspace) -> GraphInfoDto {
    let labeled_graph = workspace.projects.labeled_graph();
    extract_nodes_and_edges_from_graph(&labeled_graph)
}

/// Get a serialized representation of the dependency graph.
pub async fn dep_graph_repr(graph: &DepGraph) -> GraphInfoDto {
    let labeled_graph = graph.labeled_graph();
    extract_nodes_and_edges_from_graph(&labeled_graph)
}

pub fn respond_to_request(
    server: Server,
    tera: &mut Tera,
    graph: &GraphInfoDto,
) -> Result<(), AnyError> {
    info!(
        target: LOG_TARGET,
        r#"Starting server on "{}""#,
        server.server_addr()
    );
    for req in server.incoming_requests() {
        let response = match req.url() {
            "/graph-data" => {
                let mut response = Response::from_data(serde_json::to_string(graph)?);
                response.add_header(
                    Header::from_bytes(&b"Content-Type"[..], &b"application/json"[..]).unwrap(),
                );
                response
            }
            _ => {
                let graph_data = serde_json::to_string(graph)?;
                // Use the local version of the JS file when in development mode otherwise
                // the CDN url.
                let mut js_url = match cfg!(debug_assertions) {
                    // FIXME: We should create a separate module to store these constants
                    true => "http://localhost:5000".to_string(),
                    false => "https://cdn.com".to_string(),
                };
                js_url.push_str("/assets/index.js");
                let context = RenderContext { graph_data, js_url };
                let info = tera
                    .render_str(INDEX_HTML, &Context::from_serialize(&context)?)
                    .unwrap();
                let mut response = Response::from_data(info);
                response.add_header(
                    Header::from_bytes(&b"Content-Type"[..], &b"text/html"[..]).unwrap(),
                );
                response
            }
        };
        req.respond(response).unwrap_or_default();
    }
    Ok(())
}
