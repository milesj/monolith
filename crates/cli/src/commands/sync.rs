use crate::helpers::{build_dep_graph, create_progress_bar, load_workspace};
use moon_runner::Runner;

pub async fn sync() -> Result<(), Box<dyn std::error::Error>> {
    let done = create_progress_bar("Syncing projects...");

    let mut workspace = load_workspace().await?;
    let project_graph = workspace.generate_project_graph().await?;
    let mut project_count = 0;
    let mut dep_builder = build_dep_graph(&workspace, &project_graph);

    for project in project_graph.get_all()? {
        dep_builder.sync_project(project)?;
        project_count += 1;
    }

    let dep_graph = dep_builder.build();
    let mut runner = Runner::new(workspace);
    let results = runner.run(dep_graph, None).await?;

    if runner.has_failed() {
        done("Failed to sync projects", false);
    } else {
        done(
            format!("Successfully synced {} projects", project_count).as_ref(),
            true,
        );
    }

    runner.render_results(&results)?;

    Ok(())
}
