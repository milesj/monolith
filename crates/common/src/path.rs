pub use relative_path::*;
use starbase_styles::color;
use std::path::Path;

// Named types for better readability
pub type ProjectRelativePath = RelativePath;
pub type ProjectRelativePathBuf = RelativePathBuf;
pub type WorkspaceRelativePath = RelativePath;
pub type WorkspaceRelativePathBuf = RelativePathBuf;

#[cfg(unix)]
#[inline]
pub fn normalize_separators<T: AsRef<str>>(path: T) -> String {
    path.as_ref().replace('\\', "/")
}

#[cfg(windows)]
#[inline]
pub fn normalize_separators<T: AsRef<str>>(path: T) -> String {
    path.as_ref().replace('/', "\\")
}

#[inline]
pub fn standardize_separators<T: AsRef<str>>(path: T) -> String {
    path.as_ref().replace('\\', "/")
}

pub enum RelativeFrom<'path> {
    Project(&'path str),
    Workspace,
}

pub fn expand_to_workspace_relative<P: AsRef<str>>(
    from_format: RelativeFrom,
    path: P,
) -> WorkspaceRelativePathBuf {
    let path = standardize_separators(path.as_ref());

    match from_format {
        RelativeFrom::Project(source) => {
            // Root-level project
            if source.is_empty() || source == "." {
                WorkspaceRelativePathBuf::from(path)

                // Project-level, prefix with source path
            } else {
                let project_source = standardize_separators(source);

                if let Some(negated_glob) = path.strip_prefix('!') {
                    WorkspaceRelativePathBuf::from(format!("!{project_source}")).join(negated_glob)
                } else {
                    WorkspaceRelativePathBuf::from(project_source).join(path)
                }
            }
        }
        RelativeFrom::Workspace => WorkspaceRelativePathBuf::from(path),
    }
}

#[inline]
pub fn to_string<T: AsRef<Path>>(path: T) -> miette::Result<String> {
    let path = path.as_ref();

    match path.to_str() {
        Some(p) => Ok(p.to_owned()),
        None => Err(miette::miette!(
            "Path {} contains invalid UTF-8 characters.",
            color::path(path)
        )),
    }
}

#[inline]
pub fn to_virtual_string<T: AsRef<Path>>(path: T) -> miette::Result<String> {
    Ok(standardize_separators(to_string(path)?))
}

#[inline]
pub fn exe_name<T: AsRef<str>>(name: T) -> String {
    #[cfg(windows)]
    {
        format!("{}.exe", name.as_ref())
    }

    #[cfg(not(windows))]
    {
        name.as_ref().into()
    }
}
