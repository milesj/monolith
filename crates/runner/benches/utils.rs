use moon_contract::Platformable;
use moon_platform_node::NodePlatform;
use moon_platform_system::SystemPlatform;
use moon_project_graph::ProjectGraph;
use moon_runner::DepGraph;
use moon_task::Target;
use moon_workspace::Workspace;

pub fn setup_platforms(workspace: &mut Workspace) {
    workspace
        .projects
        .register_platform(Box::new(SystemPlatform::default()))
        .unwrap();

    workspace
        .projects
        .register_platform(Box::new(NodePlatform::default()))
        .unwrap();
}

pub fn setup_dep_graph(project_graph: &ProjectGraph) -> DepGraph {
    let mut dep_graph = DepGraph::default();

    dep_graph
        .run_target(Target::parse("base:base").unwrap(), project_graph, &None)
        .unwrap();

    dep_graph
        .run_target(
            Target::parse("depsA:dependencyOrder").unwrap(),
            project_graph,
            &None,
        )
        .unwrap();

    dep_graph
        .run_target(
            Target::parse("node:standard").unwrap(),
            project_graph,
            &None,
        )
        .unwrap();

    dep_graph
        .run_target(Target::parse("system:bash").unwrap(), project_graph, &None)
        .unwrap();

    dep_graph
        .run_target(
            Target::parse("targetScopeB:self").unwrap(),
            project_graph,
            &None,
        )
        .unwrap();

    dep_graph
}
