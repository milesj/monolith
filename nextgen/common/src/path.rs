pub use relative_path::{RelativePath, RelativePathBuf};

// Named types for better readability
pub type ProjectRelativePath = RelativePath;
pub type ProjectRelativePathBuf = RelativePathBuf;
pub type WorkspaceRelativePath = RelativePath;
pub type WorkspaceRelativePathBuf = RelativePathBuf;

#[inline]
pub fn standardize_separators<T: AsRef<str>>(path: T) -> String {
    path.as_ref().replace('\\', "/")
}
