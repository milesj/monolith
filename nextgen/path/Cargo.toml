[package]
name = "moon_path"
version = "0.1.0"
edition = "2021"

[dependencies]
relative-path = { version = "1.8.0", features = ["serde"] }
serde = { workspace = true }
