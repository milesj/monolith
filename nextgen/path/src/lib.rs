pub use relative_path::{RelativePath, RelativePathBuf};

// pub enum PathType {
//     File(String),
//     Glob(String),
// }

// pub enum Location<T> {
//     Absolute(T),
//     ProjectRelative(T),
//     WorkspaceRelative(T),
// }

// Named types for better readability
pub type ProjectRelativePath = RelativePath;
pub type ProjectRelativePathBuf = RelativePathBuf;
pub type WorkspaceRelativePath = RelativePath;
pub type WorkspaceRelativePathBuf = RelativePathBuf;

#[inline]
pub fn standardize_separators<T: AsRef<str>>(path: T) -> String {
    path.as_ref().replace('\\', "/")
}

#[inline]
pub fn expand_to_workspace_relative<F, P>(file: F, project_source: P) -> WorkspaceRelativePathBuf
where
    F: AsRef<str>,
    P: AsRef<str>,
{
    let file = file.as_ref();
    let project_source = project_source.as_ref();

    // Workspace relative negative glob
    if file.starts_with("/!") || file.starts_with("!/") {
        return WorkspaceRelativePathBuf::from(format!("!{}", standardize_separators(&file[2..])))
            .normalize();
    }

    // Workspace relative file/glob
    if let Some(path) = file.strip_prefix('/') {
        return WorkspaceRelativePathBuf::from(standardize_separators(path)).normalize();
    }

    // Project relative negative glob
    if let Some(glob) = file.strip_prefix('!') {
        return WorkspaceRelativePathBuf::from(format!("!{}", project_source))
            .join(standardize_separators(glob))
            .normalize();
    }

    // Project relative file/glob
    WorkspaceRelativePathBuf::from(project_source)
        .join(standardize_separators(file))
        .normalize()
}
