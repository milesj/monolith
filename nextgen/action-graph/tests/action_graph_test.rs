#![allow(clippy::disallowed_names)]

mod utils;

use moon_action_graph::*;
use moon_common::path::WorkspaceRelativePathBuf;
use moon_common::Id;
use moon_config::{NodeConfig, RustConfig};
use moon_node_platform::NodePlatform;
use moon_platform::PlatformManager;
use moon_platform_runtime::*;
use moon_project_graph::ProjectGraph;
use moon_rust_platform::RustPlatform;
use moon_task::{Target, Task};
use moon_test_utils2::generate_project_graph;
use proto_core::ProtoEnvironment;
use rustc_hash::FxHashSet;
use starbase_sandbox::{assert_snapshot, create_sandbox};
use std::path::Path;
use std::sync::Arc;
use utils::ActionGraphContainer;

fn create_task(id: &str, project: &str) -> Task {
    Task {
        id: Id::raw(id),
        target: Target::new(project, id).unwrap(),
        ..Task::default()
    }
}

async fn create_project_graph() -> ProjectGraph {
    generate_project_graph("projects").await
}

fn create_node_runtime() -> Runtime {
    Runtime::new(
        PlatformType::Node,
        RuntimeReq::with_version(Version::new(20, 0, 0)),
    )
}

fn create_rust_runtime() -> Runtime {
    Runtime::new(
        PlatformType::Rust,
        RuntimeReq::with_version(Version::new(1, 70, 0)),
    )
}

fn create_platform_manager(root: &Path) -> PlatformManager {
    let mut manager = PlatformManager::default();
    let proto = Arc::new(ProtoEnvironment::new_testing(root));

    manager.register(
        PlatformType::Node,
        Box::new(NodePlatform::new(
            &NodeConfig {
                version: Some(UnresolvedVersionSpec::Version(Version::new(20, 0, 0))),
                ..Default::default()
            },
            &None,
            root,
            proto.clone(),
        )),
    );

    manager.register(
        PlatformType::Rust,
        Box::new(RustPlatform::new(
            &RustConfig {
                version: Some(UnresolvedVersionSpec::Version(Version::new(1, 70, 0))),
                ..Default::default()
            },
            root,
            proto.clone(),
        )),
    );

    manager
}

fn topo(mut graph: ActionGraph) -> Vec<ActionNode> {
    let mut nodes = vec![];

    graph.reset_iterator().unwrap();

    for node in graph {
        nodes.push(node);
    }

    nodes
}

mod action_graph {
    use super::*;

    // #[test]
    // fn errors_on_cycle() {
    //     let mut graph = ProjectGraphType::new();
    //     let a = graph.add_node(create_project("a"));
    //     let b = graph.add_node(create_project("b"));
    //     graph.add_edge(a, b, DependencyScope::Build);
    //     graph.add_edge(b, a, DependencyScope::Build);

    //     let pg = ProjectGraph::new(
    //         graph,
    //         FxHashMap::from_iter([
    //             ("a".into(), ProjectNode::new(0)),
    //             ("b".into(), ProjectNode::new(1)),
    //         ]),
    //         &PathBuf::from("."),
    //     );

    //     let mut builder = ActionGraphBuilder::new(&pg).unwrap();

    //     builder.sync_project(&pg.get("a").unwrap()).unwrap();
    //     builder.sync_project(&pg.get("b").unwrap()).unwrap();

    //     builder.build().unwrap().reset_iterator().unwrap();
    // }

    mod install_deps {
        use super::*;

        #[tokio::test]
        async fn graphs() {
            let sandbox = create_sandbox("projects");
            let container = ActionGraphContainer::new(sandbox.path()).await;
            let mut builder = container.create_builder();

            let bar = container.project_graph.get("bar").unwrap();
            builder.install_deps(&bar, None).unwrap();

            let graph = builder.build().unwrap();

            assert_snapshot!(graph.to_dot());
            assert_eq!(
                topo(graph),
                vec![
                    ActionNode::SyncWorkspace,
                    ActionNode::SetupTool {
                        runtime: create_node_runtime()
                    },
                    ActionNode::InstallDeps {
                        runtime: create_node_runtime()
                    }
                ]
            );
        }

        #[tokio::test]
        async fn ignores_dupes() {
            let sandbox = create_sandbox("projects");
            let container = ActionGraphContainer::new(sandbox.path()).await;
            let mut builder = container.create_builder();

            let bar = container.project_graph.get("bar").unwrap();
            builder.install_deps(&bar, None).unwrap();
            builder.install_deps(&bar, None).unwrap();
            builder.install_deps(&bar, None).unwrap();

            let graph = builder.build().unwrap();

            assert_eq!(
                topo(graph),
                vec![
                    ActionNode::SyncWorkspace,
                    ActionNode::SetupTool {
                        runtime: create_node_runtime()
                    },
                    ActionNode::InstallDeps {
                        runtime: create_node_runtime()
                    }
                ]
            );
        }

        #[tokio::test]
        async fn installs_in_project_when_not_in_depman_workspace() {
            let sandbox = create_sandbox("dep-workspace");
            let container = ActionGraphContainer::new(sandbox.path()).await;
            let mut builder = container.create_builder();

            let inside = container.project_graph.get("in").unwrap();
            builder.install_deps(&inside, None).unwrap();

            let outside = container.project_graph.get("out").unwrap();
            builder.install_deps(&outside, None).unwrap();

            let graph = builder.build().unwrap();

            assert_snapshot!(graph.to_dot());
            assert_eq!(
                topo(graph),
                vec![
                    ActionNode::SyncWorkspace,
                    ActionNode::SetupTool {
                        runtime: create_node_runtime()
                    },
                    ActionNode::InstallProjectDeps {
                        project: Id::raw("out"),
                        runtime: create_node_runtime()
                    },
                    ActionNode::InstallDeps {
                        runtime: create_node_runtime()
                    },
                ]
            );
        }
    }

    mod run_task {
        use super::*;
        use starbase_sandbox::pretty_assertions::assert_eq;

        #[tokio::test]
        async fn graphs() {
            let sandbox = create_sandbox("projects");
            let container = ActionGraphContainer::new(sandbox.path()).await;
            let mut builder = container.create_builder();

            let project = container.project_graph.get("bar").unwrap();

            let mut task = create_task("build", "bar");
            task.platform = PlatformType::Node;

            builder.run_task(&project, &task, None).unwrap();

            let graph = builder.build().unwrap();

            assert_snapshot!(graph.to_dot());
            assert_eq!(
                topo(graph),
                vec![
                    ActionNode::SyncWorkspace,
                    ActionNode::SetupTool {
                        runtime: create_node_runtime()
                    },
                    ActionNode::SyncProject {
                        project: Id::raw("bar"),
                        runtime: create_node_runtime()
                    },
                    ActionNode::InstallDeps {
                        runtime: create_node_runtime()
                    },
                    ActionNode::RunTask {
                        interactive: false,
                        persistent: false,
                        runtime: create_node_runtime(),
                        target: task.target
                    }
                ]
            );
        }

        #[tokio::test]
        async fn ignores_dupes() {
            let sandbox = create_sandbox("projects");
            let container = ActionGraphContainer::new(sandbox.path()).await;
            let mut builder = container.create_builder();

            let project = container.project_graph.get("bar").unwrap();

            let mut task = create_task("build", "bar");
            task.platform = PlatformType::Node;

            builder.run_task(&project, &task, None).unwrap();
            builder.run_task(&project, &task, None).unwrap();
            builder.run_task(&project, &task, None).unwrap();

            let graph = builder.build().unwrap();

            assert_eq!(
                topo(graph),
                vec![
                    ActionNode::SyncWorkspace,
                    ActionNode::SetupTool {
                        runtime: create_node_runtime()
                    },
                    ActionNode::SyncProject {
                        project: Id::raw("bar"),
                        runtime: create_node_runtime()
                    },
                    ActionNode::InstallDeps {
                        runtime: create_node_runtime()
                    },
                    ActionNode::RunTask {
                        interactive: false,
                        persistent: false,
                        runtime: create_node_runtime(),
                        target: task.target
                    }
                ]
            );
        }

        #[tokio::test]
        async fn doesnt_graph_if_not_affected_by_touched_files() {
            let sandbox = create_sandbox("projects");
            let container = ActionGraphContainer::new(sandbox.path()).await;
            let mut builder = container.create_builder();

            let project = container.project_graph.get("bar").unwrap();

            let mut task = create_task("build", "bar");
            task.platform = PlatformType::Node;

            builder
                // Empty set works fine, just needs to be some
                .run_task(&project, &task, Some(&FxHashSet::default()))
                .unwrap();

            let graph = builder.build().unwrap();

            assert!(topo(graph).is_empty());
        }

        #[tokio::test]
        async fn graphs_if_affected_by_touched_files() {
            let sandbox = create_sandbox("projects");
            let container = ActionGraphContainer::new(sandbox.path()).await;
            let mut builder = container.create_builder();

            let file = WorkspaceRelativePathBuf::from("bar/file.js");

            let project = container.project_graph.get("bar").unwrap();

            let mut task = create_task("build", "bar");
            task.platform = PlatformType::Node;
            task.input_files.insert(file.clone());

            builder
                .run_task(&project, &task, Some(&FxHashSet::from_iter([file])))
                .unwrap();

            let graph = builder.build().unwrap();

            assert!(!topo(graph).is_empty());
        }

        #[tokio::test]
        async fn task_can_have_a_diff_platform_from_project() {
            let sandbox = create_sandbox("projects");
            let container = ActionGraphContainer::new(sandbox.path()).await;
            let mut builder = container.create_builder();

            // node
            let project = container.project_graph.get("bar").unwrap();

            let mut task = create_task("build", "bar");
            task.platform = PlatformType::Rust;

            builder.run_task(&project, &task, None).unwrap();

            let graph = builder.build().unwrap();

            assert_snapshot!(graph.to_dot());
            assert_eq!(
                topo(graph),
                vec![
                    ActionNode::SyncWorkspace,
                    ActionNode::SetupTool {
                        runtime: create_node_runtime()
                    },
                    ActionNode::SetupTool {
                        runtime: create_rust_runtime()
                    },
                    ActionNode::SyncProject {
                        project: Id::raw("bar"),
                        runtime: create_node_runtime()
                    },
                    ActionNode::InstallDeps {
                        runtime: create_rust_runtime()
                    },
                    ActionNode::RunTask {
                        interactive: false,
                        persistent: false,
                        runtime: create_rust_runtime(),
                        target: task.target
                    }
                ]
            );
        }

        #[tokio::test]
        async fn sets_interactive() {
            let sandbox = create_sandbox("projects");
            let container = ActionGraphContainer::new(sandbox.path()).await;
            let mut builder = container.create_builder();

            let project = container.project_graph.get("bar").unwrap();

            let mut task = create_task("build", "bar");
            task.options.interactive = true;

            builder.run_task(&project, &task, None).unwrap();

            let graph = builder.build().unwrap();

            assert_eq!(
                topo(graph).last().unwrap(),
                &ActionNode::RunTask {
                    interactive: true,
                    persistent: false,
                    runtime: Runtime::system(),
                    target: task.target
                }
            );
        }

        #[tokio::test]
        async fn sets_persistent() {
            let sandbox = create_sandbox("projects");
            let container = ActionGraphContainer::new(sandbox.path()).await;
            let mut builder = container.create_builder();

            let project = container.project_graph.get("bar").unwrap();

            let mut task = create_task("build", "bar");
            task.options.persistent = true;

            builder.run_task(&project, &task, None).unwrap();

            let graph = builder.build().unwrap();

            assert_eq!(
                topo(graph).last().unwrap(),
                &ActionNode::RunTask {
                    interactive: false,
                    persistent: true,
                    runtime: Runtime::system(),
                    target: task.target
                }
            );
        }
    }

    mod setup_tool {
        use super::*;

        #[tokio::test]
        async fn graphs() {
            let pg = ProjectGraph::default();
            let mut builder = ActionGraphBuilder::new(&pg).unwrap();
            let system = Runtime::system();
            let node = Runtime::new(
                PlatformType::Node,
                RuntimeReq::with_version(Version::new(1, 2, 3)),
            );

            builder.setup_tool(&system);
            builder.setup_tool(&node);

            let graph = builder.build().unwrap();

            assert_snapshot!(graph.to_dot());
            assert_eq!(
                topo(graph),
                vec![
                    ActionNode::SyncWorkspace,
                    ActionNode::SetupTool { runtime: node },
                    ActionNode::SetupTool { runtime: system },
                ]
            );
        }

        #[tokio::test]
        async fn graphs_same_platform() {
            let pg = ProjectGraph::default();
            let mut builder = ActionGraphBuilder::new(&pg).unwrap();

            let node1 = Runtime::new(
                PlatformType::Node,
                RuntimeReq::with_version(Version::new(1, 2, 3)),
            );
            let node2 = Runtime::new_override(
                PlatformType::Node,
                RuntimeReq::with_version(Version::new(4, 5, 6)),
            );
            let node3 = Runtime::new(PlatformType::Node, RuntimeReq::Global);

            builder.setup_tool(&node1);
            builder.setup_tool(&node2);
            builder.setup_tool(&node3);

            let graph = builder.build().unwrap();

            assert_snapshot!(graph.to_dot());
            assert_eq!(
                topo(graph),
                vec![
                    ActionNode::SyncWorkspace,
                    ActionNode::SetupTool { runtime: node3 },
                    ActionNode::SetupTool { runtime: node2 },
                    ActionNode::SetupTool { runtime: node1 },
                ]
            );
        }

        #[tokio::test]
        async fn ignores_dupes() {
            let pg = ProjectGraph::default();
            let mut builder = ActionGraphBuilder::new(&pg).unwrap();
            let system = Runtime::system();

            builder.setup_tool(&system);
            builder.setup_tool(&system);

            let graph = builder.build().unwrap();

            assert_eq!(
                topo(graph),
                vec![
                    ActionNode::SyncWorkspace,
                    ActionNode::SetupTool { runtime: system },
                ]
            );
        }
    }

    mod sync_project {
        use super::*;

        #[tokio::test]
        async fn graphs_single() {
            let pg = create_project_graph().await;
            let mut builder = ActionGraphBuilder::new(&pg).unwrap();

            let bar = pg.get("bar").unwrap();
            builder.sync_project(&bar).unwrap();

            let graph = builder.build().unwrap();

            assert_snapshot!(graph.to_dot());
            assert_eq!(
                topo(graph),
                vec![
                    ActionNode::SyncWorkspace,
                    ActionNode::SetupTool {
                        runtime: Runtime::system()
                    },
                    ActionNode::SyncProject {
                        project: Id::raw("bar"),
                        runtime: Runtime::system()
                    }
                ]
            );
        }

        #[tokio::test]
        async fn graphs_single_with_dep() {
            let pg = create_project_graph().await;
            let mut builder = ActionGraphBuilder::new(&pg).unwrap();

            let foo = pg.get("foo").unwrap();
            builder.sync_project(&foo).unwrap();

            let graph = builder.build().unwrap();

            assert_snapshot!(graph.to_dot());
            assert_eq!(
                topo(graph),
                vec![
                    ActionNode::SyncWorkspace,
                    ActionNode::SetupTool {
                        runtime: Runtime::system()
                    },
                    ActionNode::SyncProject {
                        project: Id::raw("bar"),
                        runtime: Runtime::system()
                    },
                    ActionNode::SyncProject {
                        project: Id::raw("foo"),
                        runtime: Runtime::system()
                    }
                ]
            );
        }

        #[tokio::test]
        async fn graphs_multiple() {
            let pg = create_project_graph().await;
            let mut builder = ActionGraphBuilder::new(&pg).unwrap();

            let foo = pg.get("foo").unwrap();
            builder.sync_project(&foo).unwrap();

            let bar = pg.get("bar").unwrap();
            builder.sync_project(&bar).unwrap();

            let qux = pg.get("qux").unwrap();
            builder.sync_project(&qux).unwrap();

            let graph = builder.build().unwrap();

            assert_snapshot!(graph.to_dot());
            assert_eq!(
                topo(graph),
                vec![
                    ActionNode::SyncWorkspace,
                    ActionNode::SetupTool {
                        runtime: Runtime::system()
                    },
                    ActionNode::SyncProject {
                        project: Id::raw("qux"),
                        runtime: Runtime::system()
                    },
                    ActionNode::SyncProject {
                        project: Id::raw("bar"),
                        runtime: Runtime::system()
                    },
                    ActionNode::SyncProject {
                        project: Id::raw("foo"),
                        runtime: Runtime::system()
                    },
                ]
            );
        }

        #[tokio::test]
        async fn ignores_dupes() {
            let pg = create_project_graph().await;
            let mut builder = ActionGraphBuilder::new(&pg).unwrap();

            let foo = pg.get("foo").unwrap();

            builder.sync_project(&foo).unwrap();
            builder.sync_project(&foo).unwrap();
            builder.sync_project(&foo).unwrap();

            let graph = builder.build().unwrap();

            assert_eq!(
                topo(graph),
                vec![
                    ActionNode::SyncWorkspace,
                    ActionNode::SetupTool {
                        runtime: Runtime::system()
                    },
                    ActionNode::SyncProject {
                        project: Id::raw("bar"),
                        runtime: Runtime::system()
                    },
                    ActionNode::SyncProject {
                        project: Id::raw("foo"),
                        runtime: Runtime::system()
                    }
                ]
            );
        }

        #[tokio::test]
        async fn inherits_platform_tool() {
            let pg = create_project_graph().await;
            let pm = create_platform_manager(&pg.workspace_root);
            let mut builder = ActionGraphBuilder::with_platforms(&pm, &pg).unwrap();

            let bar = pg.get("bar").unwrap();
            builder.sync_project(&bar).unwrap();

            let qux = pg.get("qux").unwrap();
            builder.sync_project(&qux).unwrap();

            let graph = builder.build().unwrap();

            assert_snapshot!(graph.to_dot());
            assert_eq!(
                topo(graph),
                vec![
                    ActionNode::SyncWorkspace,
                    ActionNode::SetupTool {
                        runtime: create_rust_runtime()
                    },
                    ActionNode::SetupTool {
                        runtime: create_node_runtime()
                    },
                    ActionNode::SyncProject {
                        project: Id::raw("qux"),
                        runtime: create_rust_runtime()
                    },
                    ActionNode::SyncProject {
                        project: Id::raw("bar"),
                        runtime: create_node_runtime()
                    }
                ]
            );
        }

        #[tokio::test]
        async fn supports_platform_override() {
            let pg = create_project_graph().await;
            let pm = create_platform_manager(&pg.workspace_root);
            let mut builder = ActionGraphBuilder::with_platforms(&pm, &pg).unwrap();

            let bar = pg.get("bar").unwrap();
            builder.sync_project(&bar).unwrap();

            let baz = pg.get("baz").unwrap();
            builder.sync_project(&baz).unwrap();

            let graph = builder.build().unwrap();

            assert_snapshot!(graph.to_dot());
            assert_eq!(
                topo(graph),
                vec![
                    ActionNode::SyncWorkspace,
                    ActionNode::SetupTool {
                        runtime: Runtime::new_override(
                            PlatformType::Node,
                            RuntimeReq::with_version(Version::new(18, 0, 0))
                        )
                    },
                    ActionNode::SetupTool {
                        runtime: create_node_runtime()
                    },
                    ActionNode::SyncProject {
                        project: Id::raw("baz"),
                        runtime: Runtime::new_override(
                            PlatformType::Node,
                            RuntimeReq::with_version(Version::new(18, 0, 0))
                        )
                    },
                    ActionNode::SyncProject {
                        project: Id::raw("bar"),
                        runtime: create_node_runtime()
                    },
                ]
            );
        }
    }

    mod sync_workspace {
        use super::*;

        #[tokio::test]
        async fn graphs() {
            let pg = ProjectGraph::default();

            let mut builder = ActionGraphBuilder::new(&pg).unwrap();
            builder.sync_workspace();

            let graph = builder.build().unwrap();

            assert_snapshot!(graph.to_dot());
            assert_eq!(topo(graph), vec![ActionNode::SyncWorkspace]);
        }

        #[tokio::test]
        async fn ignores_dupes() {
            let pg = ProjectGraph::default();

            let mut builder = ActionGraphBuilder::new(&pg).unwrap();
            builder.sync_workspace();
            builder.sync_workspace();
            builder.sync_workspace();

            let graph = builder.build().unwrap();

            assert_eq!(topo(graph), vec![ActionNode::SyncWorkspace]);
        }
    }
}
