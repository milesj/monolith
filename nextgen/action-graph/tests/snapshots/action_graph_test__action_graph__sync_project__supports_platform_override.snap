---
source: nextgen/action-graph/tests/action_graph_test.rs
expression: graph.to_dot()
---
digraph {
    0 [ label="SyncWorkspace" ]
    1 [ label="SetupNodeTool(20.0.0)" ]
    2 [ label="SyncNodeProject(bar)" ]
    3 [ label="SetupNodeTool(18.0.0)" ]
    4 [ label="SyncNodeProject(baz)" ]
    1 -> 0 [ ]
    2 -> 1 [ ]
    3 -> 0 [ ]
    4 -> 3 [ ]
}

